tests/lib/jwt/src/main.c

```diff
@@ -0,0 +1,302 @@
+ */
+#if 1
+int base64_char(int value)
+{
+	if (value < 26) {
+		return value + 'A';
+	} else if (value < 52) {
+		return value + 'a' - 26;
+	} else if (value < 62) {
+		return value + '0' - 52;
+	} else if (value == 62) {
+		return '-';
+	} else {
+		return '_';
+	}
+}
+#else
+static const char b64_table[] = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-_";
+static inline int base64_char(int value)
+{
+	return b64_table[value];
+}
+#endif
+
+struct base64_state {
+	char *buf;
+	int len;
+	unsigned char wip[3];
+	int pending;
+	bool overflowed;
+};
+
+void base64_init(struct base64_state *st, char *buf, int buflen)
+{
+	st->buf = buf;
+	st->len = buflen;
+	st->pending = 0;
+	st->overflowed = false;
+	memset(st->wip, 0, 3);
+}
+
+void base64_outch(struct base64_state *st, char ch)
+{
+	if (st->overflowed) {
+		return;
+	}
+
+	if (st->len < 2) {
+		st->overflowed = true;
+		return;
+	}
+
+	*st->buf++ = ch;
+	st->len--;
+	*st->buf = 0;
+}
+
+/* Perform a non-padded flush. */
+#if 1
+void base64_flush(struct base64_state *st)
+{
+	if (st->pending < 1) {
+		return;
+	}
+
+	base64_outch(st, base64_char(st->wip[0] >> 2));
+	base64_outch(st, base64_char(((st->wip[0] & 0x03) << 4) | (st->wip[1] >> 4)));
+
+	if (st->pending >= 2) {
+		base64_outch(st, base64_char(((st->wip[1] & 0x0f) << 2) | (st->wip[2] >> 6)));
+	}
+	if (st->pending >= 3) {
+		base64_outch(st, base64_char(st->wip[2] & 0x3f));
+	}
+
+	st->pending = 0;
+	memset(st->wip, 0, 3);
+	return;
+}
+#else
+void base64_flush(struct base64_state *st)
+{
+	if (st->pending < 1) {
+		return;
+	}
+
+	uint8_t w0 = st->wip[0];
+	uint8_t w1 = st->wip[1];
+	base64_outch(st, base64_char(w0 >> 2));
+	base64_outch(st, base64_char(((w0 & 0x03) << 4) | (w1 >> 4)));
+
+	if (st->pending < 2) {
+		goto done;
+	}
+
+	uint8_t w2 = st->wip[2];
+	base64_outch(st, base64_char(((w1 & 0x0f) << 2) | (w2 >> 6)));
+
+	if (st->pending < 3) {
+		goto done;
+	}
+
+	base64_outch(st, base64_char(w2 & 0x3f));
+
+done:
+	st->pending = 0;
+	memset(st->wip, 0, 3);
+	return;
+}
+#endif
+
+static void base64_addbyte(struct base64_state *st, uint8_t byte)
+{
+	st->wip[st->pending++] = byte;
+	if (st->pending == 3) {
+		base64_flush(st);
+	}
+}
+
+static int base64_append_bytes(const char *bytes, size_t len,
+			 void *data)
+{
+	struct base64_state *st = data;
+
+	while (len-- > 0) {
+		base64_addbyte(st, *bytes++);
+	}
+
+	return 0;
+}
+
+struct jwt_header {
+	char *typ;
+	char *alg;
+};
+
+struct json_obj_descr jwt_header_desc[] = {
+	JSON_OBJ_DESCR_PRIM(struct jwt_header, alg, JSON_TOK_STRING),
+	JSON_OBJ_DESCR_PRIM(struct jwt_header, typ, JSON_TOK_STRING),
+};
+
+struct jwt_payload {
+	s32_t exp;
+	s32_t iat;
+	char *aud;
+};
+
+struct json_obj_descr jwt_payload_desc[] = {
+	JSON_OBJ_DESCR_PRIM(struct jwt_payload, aud, JSON_TOK_STRING),
+	JSON_OBJ_DESCR_PRIM(struct jwt_payload, exp, JSON_TOK_NUMBER),
+	JSON_OBJ_DESCR_PRIM(struct jwt_payload, iat, JSON_TOK_NUMBER),
+};
+
+static void test_json_encode(void)
```

Comment:
This whole section function, and most of what is above it that it calls should just be removed. It migrated into the JWT code and will be tested as part of the JWT test. There isn't really much point in testing what is now an independent implementation.

Reply:
done